### PR TITLE
NMS-14929: Added event ingestor

### PR DESCRIPTION
**Pull request by christianpape**
* JIRA: https://issues.opennms.org/browse/NMS-14929
* JIRA: https://issues.opennms.org/browse/NMS-14928

### CODE REVIEW

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    private String category;
+    private String event;
+    private Integer id;
+    private OffsetDateTime eventTime;
+    private String edgeName;
+    private String enterpriseUsername;
+    private String message;
+    private String severity;
+
+    private EnterpriseEvent(final Builder builder) {
```
Comment (aramosvizcarra) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java:20]: Is there any of this fields required to be not null?

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    private OffsetDateTime eventTime;
+    private String proxyUsername;
+    private String networkName;
+    private String enterpriseName;
+    private String gatewayName;
+    private String message;
+    private String severity;
+
+    private ProxyEvent(final ProxyEvent.Builder builder) {
+        this.detail = builder.detail;
```
Comment (aramosvizcarra) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java:22]: same here, are any of this fields required to be not null?

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
+package org.opennms.velocloud.client.api.model;
```
Comment (aramosvizcarra) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java:1]: Missing license

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
+package org.opennms.velocloud.client.api.model;
```
Comment (aramosvizcarra) [inline client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java:1]: Missing license

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventConfExtension.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package org.opennms.velocloud.events;
```
Comment (aramosvizcarra) [inline plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventConfExtension.java:1]: Missing license

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 6]
```diff
@@ -0,0 +1,320 @@
+package org.opennms.velocloud.events;
```
Comment (aramosvizcarra) [inline plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:1]: missing license

**Review by aramosvizcarra — CHANGES_REQUESTED** [review]
a couple of questions an a few license missing. Otherwise looks good to me.

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java (hunk 1 above)
Comment (christianpape) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java:49]: done

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java (hunk 2 above)
Comment (christianpape) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java]: done

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java (hunk 3 above)
Comment (christianpape) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java:29]: done

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java (hunk 4 above)
Comment (christianpape) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java:29]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventConfExtension.java (hunk 5 above)
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventConfExtension.java:29]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java (hunk 6 above)
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:29]: done

File: client/v2/src/build/resources/vmware-sdwan-api-doc.yaml.orig [hunk 7]
```diff
@@ -6,7 +6,7 @@ info:
...
...eployed, number of Customer events generated, etc.). Alongside a `data` array which contains a (possibly truncated) result set, paginated responses include a `metaData` object. In the event that a query matches a result set that is larger than the server-enforced maximum size (or a user-specified `limit`, if one was provided), a `nextPageLink` token is included in the response `metaData` object. That token may be passed to the server in a subsequent GET request as a query parameter called `nextPageLink` in order to fetch the next page in the result set.</p>             <p style="font-weight: 100;">For example, suppose a Customer has 3000 Edges and an API user with access to that Customer makes an initial query to `GET /enterprises/<enterpriseLogicalId>/edges`. We expect that this query will "match" all 3000 Customer Edges. The server, however, enforces a maximum result set size of 2048, so it produces a `data` array containing 2048 Edges, and a `metaData` object alongside that array which includes a `nextPageLink` token. The user would fetch the next page in the result set by making a request to: `GET /enterprises/<enterpriseLogicalId>/edges?nextPageLink=<nextPageToken>`.</p>     '
```
Comment (aramosvizcarra) [inline client/v2/src/build/resources/vmware-sdwan-api-doc.yaml.orig:9]: Not sure if you should be doing this changes in the client/v2/src/build/resources/vmware-sdwan-api-doc.yaml instead? the **vmware-sdwan-api-doc.yaml.orig** file was meant to keep the original openapi descriptor, and the **vmware-sdwan-api-doc.yaml**  was meant for use to change and also is the one used to generate the api generated source.

**Review by aramosvizcarra — CHANGES_REQUESTED** [review]
I missed a question about the vmware-sdwan-api-doc.yaml.orig file. Otherwise looks good.

**Review by aramosvizcarra — APPROVED** [review]
LGTM

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+    private String getMetadataForKey(final String key, final List<MetaData> metaDataList) {
+        for (final MetaData metaData : metaDataList) {
+            if (AbstractRequisitionProvider.VELOCLOUD_METADATA_CONTEXT.equals(metaData.getContext()) && key.equals(metaData.getKey())) {
+                return metaData.getValue();
+            }
+        }
+        return null;
+    }
+
+    private String parseJson(String detail, String key) {
```
Comment (fooker) [inline plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:298]: Please add a comment why this is as ugly as this.

**Review by fooker — CHANGES_REQUESTED** [review]

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        this.detail = Objects.requireNonNull(builder.detail);
+        this.category = Objects.requireNonNull(builder.category);
+        this.event = Objects.requireNonNull(builder.event);
+        this.id = Objects.requireNonNull(builder.id);
+        this.eventTime = Objects.requireNonNull(builder.eventTime);
+        this.message = Objects.requireNonNull(builder.message);
+        this.severity = Objects.requireNonNull(builder.severity);
+
+        this.edgeName = builder.edgeName;
+        this.enterpriseUsername = builder.enterpriseUsername;
```
[the remark this answers is not in the document]
Comment (christianpape) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java:59]: yes

File: client/v2/src/build/resources/vmware-sdwan-api-doc.yaml.orig (hunk 7 above)
Comment (christianpape) [reply client/v2/src/build/resources/vmware-sdwan-api-doc.yaml.orig]: done

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+public class EnterpriseEvent {
+    private String detail;
+    private String category;
+    private String event;
+    private Integer id;
+    private OffsetDateTime eventTime;
+    private String edgeName;
+    private String enterpriseUsername;
+    private String message;
+    private String severity;
```
[the remark this answers is not in the document]
Comment (christianpape) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java:47]: done

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                .add("detail:" + this.detail)
+                .add("category:" + this.category)
+                .add("event:" + this.event)
+                .add("id:" + this.id)
+                .add("eventTime:" + this.eventTime)
+                .add("edgeName:" + this.edgeName)
+                .add("enterpriseUsername:" + this.enterpriseUsername)
+                .add("message:" + this.message)
+                .add("severity:" + this.severity)
+                .toString();
```
[the remark this answers is not in the document]
Comment (christianpape) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/EnterpriseEvent.java:175]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+    public static final String ENTERPRISE_EVENTS_UEI = "uei.opennms.org/vendor/vmware/velocloud/enterpriseEvents";
+    public static final String PROXY_EVENTS_UEI = "uei.opennms.org/vendor/vmware/velocloud/proxyEvents";
+    private final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    private ScheduledFuture<?> scheduledFuture;
+    private long delay = 0L;
+    private final ClientManager clientManager;
+    private final ConnectionManager connectionManager;
+    private final NodeDao nodeDao;
+    private final EventForwarder eventForwarder;
+    private final static Map<String, Severity> SEVERITY_MAP = new TreeMap<>();
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:87]: done

[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:87]: done

File: client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    private String category;
+    private String event;
+    private Integer id;
+    private OffsetDateTime eventTime;
+    private String proxyUsername;
+    private String networkName;
+    private String enterpriseName;
+    private String gatewayName;
+    private String message;
+    private String severity;
```
[the remark this answers is not in the document]
Comment (christianpape) [reply client/api/src/main/java/org/opennms/velocloud/client/api/model/ProxyEvent.java:48]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    public VelocloudEventIngestor(final ClientManager clientManager,
+                                  final ConnectionManager connectionManager,
+                                  final NodeDao nodeDao,
+                                  final EventForwarder eventForwarder) {
+        this.clientManager = clientManager;
+        this.connectionManager = connectionManager;
+        this.nodeDao = nodeDao;
+        this.eventForwarder = eventForwarder;
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:109]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+        for (final String alias : connectionManager.getAliases()) {
+            final Optional<Connection> connection = connectionManager.getConnection(alias);
+
+            if (connection.isPresent()) {
+                VelocloudApiPartnerClient velocloudApiPartnerClient = null;
+                try {
+                    velocloudApiPartnerClient = clientManager.getPartnerClient(VelocloudApiClientCredentials.builder()
+                            .withOrchestratorUrl(connection.get().getOrchestratorUrl())
+                            .withApiKey(connection.get().getApiKey())
+                            .build());
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:142]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+                } catch (Exception e) {
+                    LOG.debug("Cannot create partner connection for the given credentials.");
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:144]: removed

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+                }
+
+                if (velocloudApiPartnerClient != null) {
+                    try {
+                        final String foreignSource = String.format("%s-%s", AbstractRequisitionProvider.VELOCLOUD_PARNTER_IDENTIFIER, connection.get().getAlias());
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:149]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+        } else {
+            builder.addParameter(ImmutableEventParameter.newInstance("eventTime", null));
+        }
+
+        eventForwarder.sendSync(builder.build());
+    }
+
+    private void processCustomerEvents(final Instant start, final Instant end, final String foreignSource, final VelocloudApiCustomerClient velocloudApiCustomerClient) throws VelocloudApiException {
+        final List<EnterpriseEvent> enterpriseEventList = velocloudApiCustomerClient.getEnterpriseEvents(start, end);
+        LOG.debug(enterpriseEventList.size() + " enterprise events found.");
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:266]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+    }
+
+    private void processPartnerEvents(final Instant start, final Instant end, final String foreignSource, final VelocloudApiPartnerClient velocloudApiPartnerClient) throws VelocloudApiException {
+        final List<ProxyEvent> enterpriseEventList = velocloudApiPartnerClient.getProxyEvents(start, end);
+        LOG.debug(enterpriseEventList.size() + " proxy events found.");
+        for (final ProxyEvent proxyEvent : enterpriseEventList) {
+            if (!Strings.isNullOrEmpty(proxyEvent.getGatewayName())) {
+                processEvent(foreignSource, proxyEvent.getGatewayName(), proxyEvent);
+            } else {
+                LOG.debug("ignoring proxy event #" + proxyEvent.getId() + " due to missing gatewayName.");
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:284]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+                processEvent(foreignSource, enterpriseEvent.getEdgeName(), enterpriseEvent);
+            } else {
+                LOG.debug("ignoring enterprise event #" + enterpriseEvent.getId() + " due to missing edgeName.");
+            }
+        }
+    }
+
+    private void processPartnerEvents(final Instant start, final Instant end, final String foreignSource, final VelocloudApiPartnerClient velocloudApiPartnerClient) throws VelocloudApiException {
+        final List<ProxyEvent> enterpriseEventList = velocloudApiPartnerClient.getProxyEvents(start, end);
+        LOG.debug(enterpriseEventList.size() + " proxy events found.");
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:279]: done

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java (hunk 8 above)
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:298]: Replaced with Yaml parser

File: plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 21]
```diff
@@ -1,7 +1,22 @@
...
-           xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0 https://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd">
+           xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0"
+           xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0
+		http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd
+
+		http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0
+		http://aries.apache.org/schemas/blueprint-cm/blueprint-cm-1.1.0.xsd">
+
+
+<reference id="clientProvider" interface="org.opennms.velocloud.client.api.VelocloudApiClientProvider"/>
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml:10]: done

File: plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 22]
```diff
@@ -32,6 +47,20 @@
     </bean>
     <service ref="customerRequisitionProvider" interface="org.opennms.integration.api.v1.requisition.RequisitionProvider"/>
 
+    <bean id="velocloudEventService" class="org.opennms.velocloud.events.VelocloudEventIngestor" destroy-method="destroy">
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml:50]: Initialization and restarts are done by the setDelay method.

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 23]
```diff
@@ -0,0 +1,348 @@
...
+                        final String foreignSource = String.format("%s-%s", AbstractRequisitionProvider.VELOCLOUD_CUSTOMER_IDENTIFIER, connection.get().getAlias());
+                        processCustomerEvents(lastPoll, now, foreignSource, velocloudApiCustomerClient);
+                    } catch (Exception e) {
+                        LOG.error("Error processing customer events.", e);
+                    }
+                }
+            }
+        }
+
+        lastPoll = now.plus(1, ChronoUnit.MILLIS);
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:178]: It's inclusive

File: plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java [hunk 24]
```diff
@@ -0,0 +1,348 @@
...
+        LOG.debug("Velocloud Event Ingestor is polling for events.");
+
+        final Instant now = Instant.now();
+
+        if (lastPoll == null) {
+            lastPoll = now.minus(delay, ChronoUnit.MILLIS);
+        }
+
+        for (final String alias : connectionManager.getAliases()) {
+            final Optional<Connection> connection = connectionManager.getConnection(alias);
```
[the remark this answers is not in the document]
Comment (christianpape) [reply plugin/src/main/java/org/opennms/velocloud/events/VelocloudEventIngestor.java:134]: Removed

File: plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 25]
```diff
@@ -1,7 +1,22 @@
 <blueprint xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-           xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0 https://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd">
+           xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0"
+           xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0
+		http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd
+
+		http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0
+		http://aries.apache.org/schemas/blueprint-cm/blueprint-cm-1.1.0.xsd">
```
Comment (fooker) [inline plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml:7]: Strange formatting

**Review by fooker — CHANGES_REQUESTED** [review]

File: plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 25 above)
Comment (christianpape) [reply plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: done

**Review by fooker — APPROVED** [review]

**Pull request merged**